06_gpu_and_ml/audio-to-text/parakeet.py

```diff
@@ -0,0 +1,246 @@
+# - The `web` method creates a FastAPI app using [`modal.asgi_app`](https://modal.com/docs/reference/modal.asgi_app#modalasgi_app) that serves a
+# [websocket](https://modal.com/docs/guide/webhooks#websockets) endpoint for real-time audio transcription.
+
+# Additionally, since this app is running in its own container, we make a `.remote` call to invoke
+# the `transcribe` method. This allows us to use the GPU for inference.
+
+class_name = "parakeet"
+
+
+@server_app.cls(volumes={"/cache": model_cache}, gpu="a10g", image=image)
+class Parakeet:
+    @modal.enter()
+    def load(self):
+        self.model = nemo_asr.models.ASRModel.from_pretrained(
+            model_name="nvidia/parakeet-tdt-0.6b-v2"
+        )
+
+    @modal.method()
+    def transcribe(self, audio_bytes: bytes) -> str:
+        import numpy as np
+
+        audio_data = np.frombuffer(audio_bytes, dtype=np.int16)
+        output = self.model.transcribe([audio_data])
+        return output[0].text
+
+    @modal.asgi_app()
+    def web(self):
+        from fastapi import FastAPI, Response, WebSocket, WebSocketDisconnect
+
+        web_app = FastAPI()
+
+        @web_app.get("/status")
+        async def status():
+            return Response(status_code=200)
+
+        @web_app.websocket("/ws")
+        async def websocket_endpoint(ws: WebSocket):
+            await ws.accept()
+            buffer = bytearray()
+
+            try:
+                while True:
+                    chunk = await ws.receive_bytes()
+                    buffer.extend(chunk)
+
+                    if len(buffer) > BUFFER_SIZE:
+                        audio_bytes = bytes(buffer)
+                        buffer.clear()
+
+                        try:
+                            text = self.transcribe.remote(audio_bytes)
+                            await ws.send_text(text)
+                        except Exception as e:
+                            print("❌ Transcription error:", e)
+                            await ws.close(code=1011, reason="Internal server error")
+            except WebSocketDisconnect:
+                print("WebSocket disconnected")
+
+        return web_app
+
+
+# ## Client
+# Next, let's test the model with a simple client that streams audio data to the server, and prints
+# out the transcriptions in real-time to our terminal. We can also run this using Modal!
+
+# ## Image
+# Using a secondary image for the client allows us to keep the client dependencies separate from the server dependencies.
+# We'll use  python's `websockets` library to create a websocket client
+# that sends audio data to the server and receives transcriptions in real-time.
+
+
+client_image = (
+    modal.Image.debian_slim(python_version="3.10")
+    .apt_install("portaudio19-dev")
+    .pip_install("websockets==15.0.1", "sounddevice==0.5.1", "numpy==2.2.5")
+)
+client_app = modal.App("parakeet-client", image=client_image)
```

Comment:
Why two apps?

Reply:
not sure how to do this with one app - we need to modal serve the server and run the client locally - taking suggestions?